### PR TITLE
Use char-wise width instead of `str`-width

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR updates various references in the linter to compute the line-width for summing the width of each `char` in a `str` instead of computing the width of the `str` itself.

Refer to #12133 for more details.

fixes: #12130 

## Test Plan

Add a file with null (`\0`) character which is zero-width. Run this test case on `main` to make sure it panics and switch over to this branch to make sure it doesn't panic now.

### CODE REVIEW

File: crates/ruff_linter/src/rules/pycodestyle/snapshots/ruff_linter__rules__pycodestyle__tests__E501_E501_4.py.snap
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pycodestyle/snapshots/ruff_linter__rules__pycodestyle__tests__E501_E501_4.py.snap:1]: Hmm, why is this considered to be a binary file?

File: crates/ruff_linter/src/fix/snippet.rs [hunk 1]
```diff
@@ -35,6 +35,11 @@ impl SourceCodeSnippet {
     /// Returns `true` if the source code should be truncated when included in a user-facing
     /// diagnostic.
     fn should_truncate(source_code: &str) -> bool {
-        source_code.width() > 50 || source_code.contains(['\r', '\n'])
+        source_code
+            .chars()
+            .map(|c| c.width().unwrap_or(0))
+            .sum::<usize>()
+            > 50
```
Comment (MichaReiser) [inline crates/ruff_linter/src/fix/snippet.rs:42]: I think we can keep using `width` here as it isn't related to any "overlong" logic.

**Review by MichaReiser — APPROVED** [review]

File: crates/ruff_linter/src/rules/pycodestyle/snapshots/ruff_linter__rules__pycodestyle__tests__E501_E501_4.py.snap
Comment (dhruvmanila) [reply crates/ruff_linter/src/rules/pycodestyle/snapshots/ruff_linter__rules__pycodestyle__tests__E501_E501_4.py.snap:1]: It contains `\0` character. I think I'll just inline the source code of that file as a separate test case to not do this.

Comment (dhruvmanila) [reply crates/ruff_linter/src/rules/pycodestyle/snapshots/ruff_linter__rules__pycodestyle__tests__E501_E501_4.py.snap:1]: I think it's just the diff tab that renders it like that. The normal GitHub view is as: https://github.com/astral-sh/ruff/blob/dhruv/char-wise-width/crates/ruff_linter/resources/test/fixtures/pycodestyle/E501_4.py

**Pull request merged**